solution/backend/resources/v3serializers/resources.py

```diff
@@ -141,28 +143,73 @@ def validate_doc_type(self, value):
-        instance.document_number = validated_data.get('document_number', instance.document_number)
-        instance.date = validated_data.get('date', instance.date)
-        instance.approved = validated_data.get('approved', instance.approved)
-        instance.doc_type = validated_data.get('doc_type', instance.doc_type)
-        instance.category = self.get_category()
+        created = self.context.get("created", False)
+
+        # set basic fields if instance is new
+        if created:
+            instance.url = validated_data.get('url', instance.url)
+            instance.description = validated_data.get('description', instance.description)
+            instance.name = validated_data.get('name', instance.name)
+            instance.docket_numbers = validated_data.get('docket_numbers', instance.docket_numbers)
+            instance.document_number = validated_data.get('document_number', instance.document_number)
+            instance.date = validated_data.get('date', instance.date)
+            instance.approved = validated_data.get('approved', instance.approved)
+            instance.doc_type = validated_data.get('doc_type', instance.doc_type)
+            instance.category = self.get_category()
 
         # set the locations on the instance
+        self.set_locations(instance, validated_data["sections"], validated_data["section_ranges"])
+
+        # reapply changelog if instance is not new
+        if not created:
+            self.apply_changelog(instance)
+
+        # set the group on the instance if instance is new
+        if created:
+            self.set_group(instance)
+
+        # save and return
+        instance.save()
+        return instance
+
+    def get_location_objects(self, locations):
+        queries = []
+        for i in locations:
+            t = i["type"]
+            queries.append(Q(title=i["title"]) & Q(part=i["part"]) & Q(**{f"{t}__{t}_id": i[f"{t}_id"]}))
+        if not queries:
+            return []
+        q = queries[0]
+        for i in queries[1:]:
+            q |= i
+        return AbstractLocation.objects.filter(q).select_subclasses()
+
+    def apply_changelog(self, instance):
+        all_additions = []
+        all_removals = []
+        for change in instance.location_history:
+            additions = [i.copy() for i in change["additions"] + change["bulk_adds"]]
+            removals = [i.copy() for i in change["removals"]]
+            for i in additions + removals:
+                del i["id"]
+            for i in additions:
+                all_removals.remove(i) if i in all_removals else all_additions.append(i)
+            for i in removals:
+                all_additions.remove(i) if i in all_additions else all_removals.append(i)
+        [instance.locations.remove(i) for i in self.get_location_objects(all_removals)]
+        [instance.locations.add(i) for i in self.get_location_objects(all_additions)]
```

Comment:
I realize this code looks like it could be simplified by just applying the changelog entries one at a time using `instance.locations.remove(...)` and `add(...)`. However this method is guaranteed to only need 2 queries per FR doc to run, instead of one DB query _per location object added or removed_.